pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -426,17 +428,25 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
       return response;
     }
 
-    final OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
+    OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
 
     long minionExpectedCrc = -1L;
     boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
-    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    String userAgentHeader = headers.getHeaderString(HttpHeaders.USER_AGENT);
+    boolean minionHeaderPresent = userAgentHeader.contains(CommonConstants.Minion.MINION_HEADER_PREFIX);
+    String taskType = UserAgentHeader.getTaskType(userAgentHeader);
     boolean metadataRewritten = false;
+
     if (isMinionConfigured) {
```

Comment:
why have we introduced this block of code based on the existence of minion task?

Reply:
If minion is configured in the table config, that means that a race condition between minion and minion uploads or a race condition between hadoop and minion uploads can happen. We want to write to zk as little as possible, and handling the race condition only in minion configured tasks allows us to cut down writes a little since part of handling the race condition is an extra write to zk to write segmentPushStatus.

Reply:
why should that logic be handled or even be specific to minion. Should we not simply have a compare and set api? Minion can read the crc or some unique identifier (zk segment metadata node version) when it downloads and upload api can take this as input and do the comparison.

Reply:
We discussed this alternative, and arrived at the conclusion that we don't want to add extra zk writes for use  cases that do not need it. We can talk about this.